File: IDP - Dynamic Mapping/Assets/Scripts/GeometryClustering/Circle.cs
using UnityEngine;

public class Circle : Primitive {
    public Color circleColor = Color.red;

    // Center and radius of the circle:
    private float xc, yc, R;
    private float xcP, ycP;     // Derivative of xc and yc (used to estimate the speed of the circle)

    public Vector2 position { get => new Vector2(xc, yc); }

    public Circle(float xc, float yc, float R) {
        this.xc = xc;
        this.yc = yc;
        this.R = R;

        this.xcP = this.ycP = 0;
    }

    public void DrawGizmos() {
        Gizmos.color = circleColor;
        Gizmos.DrawSphere(new Vector3(xc, 0.2f, yc), R);
    }

    // Supposing that this circle belongs to the current model of the environment, use the
    // given circle (that is supposed to be matched with this one) to update the center and
    // radius of this circle:
    public void UpdateCircleUsingMatching(Circle other) {
        float newXc = (xc + other.xc) / 2;
        float newYc = (yc + other.yc) / 2;
        R = (R + other.R) / 2;

        // Use the newXc and newYc to update the circle speed estimate, using a simple
        // exponential low pass filter:
        const float m = 0.9f;
        other.xcP = xcP = m * xcP + (1 - m) * (newXc - xc);
        other.ycP = ycP = m * ycP + (1 - m) * (newYc - yc);

        // Update xc and yc:
        xc = newXc; yc = newYc;
    }

    // Compute the Euclidean distance between the two circles centers:
    public float DistanceFrom(Circle other) {
        float dX = other.xc - xc;
        float dY = other.yc - yc;

        return Mathf.Sqrt(dX * dX + dY * dY);
    }

    // All the points belonging to a circle have the same speed, which is the speed
    // of the center of the circle:
    public Vector2 VelocityOfPoint(float x, float y) {
        return new Vector2(xcP, ycP);
    }
}

File: DynamicMapping/mapping/geometry_maps/circle.py
import numpy as np

class Circle:

    def __init__(self, xc, yc, R):

        # Center and radius of the circle:
        self.xc = xc
        self.yc = yc
        self.R = R

        # Derivative of xc and yc (used to estimate the speed of the circle):
        self.d_xc = 0
        self.d_yc = 0

        # If the circle is consistent with the current observations:
        self.is_valid = True

        # Color of the circle (for drawing):
        self.circle_color = (255, 0, 0)


    def draw(self, scene : 'Scene'):
        scene.draw_circle(self.xc, self.yc, self.R, self.circle_color)

    # Supposing that this circle belongs to the current model of the environment, use the
    # given circle (that is supposed to be matched with this one) to update the center and
    # radius of this circle:
    def update_circle_using_matching(self, other):
        new_xc = (self.xc + other.xc) / 2
        new_yc = (self.yc + other.yc) / 2
        self.R = (self.R + other.R) / 2

        # Use the new_xc and new_yc to update the circle speed estimate, using a simple
        # exponential low pass filter (TODO: Correct this):
        m = 0.95
        other.d_xc = self.d_xc = m * self.d_xc + (1 - m) * (new_xc - xc)
        other.d_yc = self.d_yc = m * self.d_yc + (1 - m) * (new_yc - yc)

        # Update xc and yc:
        xc = new_xc
        yc = new_yc

    # Compute the Euclidean distance between the two circles centers:
    def distance_from(self, other : 'Circle'):
        dX = other.xc - self.xc
        dY = other.yc - self.yc

        return np.Sqrt(dX * dX + dY * dY)

    # All the points belonging to a circle have the same speed, which is the speed
    # of the center of the circle:
    def velocity_of_point(self, x, y):
        return self.d_xc, self.d_yc